CareEvolution/ResearchKit
# Add CEV[1]RangeOfMotionStep

Closes https://github.com/CareEvolution/CEVResearchKit/issues/285. 

In essence, the RangeOfMotion steps use a screen tap to toggle on a device motion recorder which detects high-resolution rotation data until a second tap completes the step. As a convenience, in addition to providing the raw recorder data as a separate file, the RangeOfMotion steps provide angles of flexion and extension, which approximate the minimum and maximum of the joint motion.

In reviewing the differences between the existing ORK[1]RangeOfMotionStep (for the knee) and ORK[1]ShoulderRangeOfMotionSteps to allow deserialization of the steps directly for customization, I noticed that they both could be replicated into a more flexible RangeOfMotionStep if some elements were decoupled. A goal of this implementation is that it requires no changes on the server as it returns an ORK[1]RangeOfMotionResult which is already handled for data persistence and export.

Another improvement of this implementation is that it also bundles the device motion recorder’s output and adds it to the result. Currently, in ORK[1]RangeOfMotionStep, only the calculated angles are returned. Note that for these to be exportable, Consumers will need to be modified to accept/process the ORK[1]FileResult now being sent.

![iPhoneAxes](https://github.com/CareEvolution/ResearchKit/assets/1008462/3806be9a-d4c0-4850-b30a-01a37c1d4643)

Considering there are 3 axes of motion on an iPhone and a range of motion test is either flexion or extension of a targeted muscle/muscle group, the measurement could be further simplified/defined as:
- `motionType`: flexion or extension
- `axisOfRotation`: pitch, roll, or yaw
- `measuredDirectionOfRotation`: clockwise or counter-clockwise
- `offsetFromReferenceAxis`: where to measure zero from in relation to the range of motion test

Some examples of configuration to measure joints:

 - Flexure of shoulder:
    - iPhone in Upside Down orientation on lateral aspect of deltoid, held by alternate hand, alternatively can hold iPhone Upside Down orientation in right hand such that right index finger can tap screen to begin/end test
    - CEV1MotionType.flexion
    - CEV1AxisOfRotation.pitch
    - CEV1DirectionOfRotation.clockwise
    - offsetFromReferenceAxis = 270.0

 - Extension of knee:
    - iPhone in Upside Down orientation on medial aspect of shin, right below knee
    - CEV1MotionType.extension
    - CEV1AxisOfRotation.pitch
    - CEV1DirectionOfRotation.clockwise
    - offsetFromReferenceAxis = 270.0
 
 - Flexure of knee:
    - iPhone in Upside Down orientation on medial aspect of shin, right below knee
    - CEV1MotionType.flexion
    - CEV1AxisOfRotation.pitch
    - CEV1DirectionOfRotation.counterClockwise
    - offsetFromReferenceAxis = 0

This step is also an exercise in Swift development on ResearchKit which could enable more rapid prototype deployment in the future. Some existing Objective-C classes needed further exposure to be seen by Swift. Since there are some nuances of using NSCoding and Equivalence on a subclass of an NSObject in Swift, tests for NSCoding and equivalence were added to ensure the new Swift classes behave as expected.

## Testing

To test two use cases, Flexure of the Shoulder and Extension of the Knee, run the two surveys below (open the disclosure triangle to expand JSON), following the instructions contained in the InstructionStep preceding the RangeOfMotionStep in each survey. Note that since the server doesn’t currently recognize the TouchAnywhereStep and the RangeOfMotionStep, you will have to force publish the surveys. Then run an export, ensuring CSV SurveyResults are selected, and validate the following:
- locate the SurveyRangeOfMotionResults_… file in the export:
    - note that you see both results (shoulder and knee)
    - note that extended/flexed numbers match what you would expect
        - knee: normal extension of the knee would be from ~0 degrees flexed to ~ 90 degrees extended
        - shoulder: normal flexure of the shoulder would be from ~0 degrees extended to ~180 degrees flexed (raised)
- open the PPT in Raw Data Viewer (tap on icon at left in list view)
    - click on the Reports tab 
        - verify deviceMotion recorder files exist for each completed task _(NOTE: these will not appear in the export until some development is done in Consumers to link these to this task)_

<details>
  <summary>Flexure of Shoulder</summary>

```json
{
  "type": "RKStudioOrderedTask",
  "name": "Survey",
  "identifier": "Survey",
  "steps": [
      {
        "type": "InstructionStep",
        "text": "This test will measure the angle of flexure of a shoulder.",
        "detailText": "With your arm at your side, place the phone in your hand such that it is upside-down and face out and your index finger can tap the screen. Wait for the spoken instructions. When instructed, you will lift that arm as high as you can, exercising the range of motion of your shoulder. At the end of the flexure, tap the screen again.",
        "title": "Shoulder Flexure",
        "identifier": "Instruction"
      },
      {
        "type": "TouchAnywhereStep",
        "identifier": "TouchAnywhere",
        "instructionText": "Assume starting posture as instructed and touch anywhere to start the test."
      },
      {
        "type": "RangeOfMotionStep",
        "identifier": "ShoulderFlexion",
        "motionType": "Flexion",
        "axisOfRotation": "Pitch",
        "measuredDirectionOfRotation": "Clockwise",
        "offsetFromReferenceAxisDegrees": 270,
        "spokenInstruction": "Flex your shoulder, raising your arm as high or far as you can, then tap the screen again."
      }
  ]
}
```
</details>

<details>
  <summary>Extension of Knee (RK version 2)</summary>

```json
{
  "type": "RKStudioOrderedTask",
  "name": "Survey",
  "identifier": "Survey",
  "rkVersion": 2,
  "steps": [
      {
        "type": "InstructionStep",
        "text": "This test will measure the angle of extenion of a knee.",
        "detailText": "While seated with feet on the floor, place the phone on the medial aspect of the shin just below knee such that it is upside-down and face out and your index finger can tap the screen. Wait for the spoken instructions. When instructed, you will extend that leg as far as you can, exercising the range of motion of your knee. At the end of the extension, tap the screen again.",
        "title": "Knee Extension",
        "identifier": "Instruction"
      },
      {
        "type": "TouchAnywhereStep",
        "identifier": "TouchAnywhere",
        "instructionText": "Assume starting posture as instructed and touch anywhere to start the test."
      },
      {
        "type": "RangeOfMotionStep",
        "identifier": "ShoulderFlexion",
        "motionType": "Extension",
        "axisOfRotation": "Pitch",
        "measuredDirectionOfRotation": "Clockwise",
        "offsetFromReferenceAxisDegrees": 270,
        "spokenInstruction": "Extend your knee, raising your leg as far as you can, then tap the screen again."
      }
  ]
}
```
</details>

cc: @Suvarna-Thejas-CE , @vladtar
